# Contents

* [Reproducing Table 1 - The Impact of On-Base and Slugging Percentage on Winning](#Reproducing-Table-1-The-Impact-of-On-Base and-Slugging-Percentage-on-Winning)

* **Basis of Moneyball**:
    * Before Billy Beane, in baseball most of the scouts and teams measured a batter and the hitting ability by statistics like Slugging Percentage (how hard you can hit). None of these teams valued the player's ability to draw a walk (after 4 balls).
    * Billy Beane focussed more on on-base percentage, that took into consideration the players ability to walk, since ideally it doesn't make a difference if the player hits or walks to first base. Since drawing a walk wasn't considered an important skill by other players - these player's were very cheap to buy!

* **Slugging Percentage(SLG):**
    * **SLG = (Singles + 2 * Doubles + 3 * Triples + 4 * HomeRuns) / At bats**
    * Single = 1b. Double = 2b. Triple = 3b. HomeRun = hr. At bats = ab.
    * Measure of the batting productivity of a hitter.
    * SLG = Total bases (number of bases a player has gained with hits) / At bats (batters turn batting against a pitcher)
    * Unlike on-base percentage, slugging percentage deals only with hits and **does not include walks** and hit-by-pitches in its equation.
    * Slugging percentage differs from batting average in that all hits are not valued equally.


* **On-Base Percentage:**
    * **OBP = (Hits + Walks + Hit by Pitch) / (All bats + Hit by Pitch + Sacrifice Flies)**
    * hit = h. walk = bb. Hit by Pitch = hbp. Sacrifice Flies = sf.
    * Sacrifice Plies = when a batter hits the ball and is caught, but in the process a runner on base scores a run.

## Analyzing the Impact of On Base and Slugging Percentage on Winning

* Uses **regression analysis:**
    * mechanism for establishing the relationship between a variable and another group of variables.


* Show that on base percentage is more statistically significant in determing success, as compared to slugging percentage.


* **Challenges & Solutions:**
    * The success rate / win rate, will depend not just on the on-base & slugging percentage of a team, but also on the performance of their opponent
        * Sol - First calculate the on-base & slugging percentage of each team, across the season. Then work out the on-base and slugging percentage statistics against each team (success of opponents agains the team, across the season). Run regressions on this wherein the target variable is the win percentage.
        * Reg1 --> Looks at onbase percentage and onbase percentage against for each team.
        * Reg2 --> Looks at the slugging percentage and slugging percentage aginst for each team.
        * Reg3 --> Include both onbase and slugging percentage, for and against for each team.
        * Reg4 --> Same as 3 but imposes that the the magnitude should be equal for - for & against of each percentage.

In [6]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [7]:
Teams = pd.read_excel("datasets/Game logs 1999-2003.xlsx")

In [8]:
print(Teams.columns.tolist())

['date', 'game_number', 'day', 'visitor', 'visitor_lg', 'visitor_game_number', 'home', 'home_lg', 'home_game_number', 'visitor_score', 'home_score', 'outs', 'daynight', 'completion', 'forfeit', 'protest', 'park', 'attendance', 'game_minutes', 'visitor_linescore', 'home_linescore', 'visitor_ab', 'visitor_h', 'visitor_2b', 'visitor_3b', 'visitor_hr', 'visitor_rbi', 'visitor_sh', 'visitor_sf', 'visitor_hbp', 'visitor_bb', 'visitor_ibb', 'visitor_so', 'visitor_sb', 'visitor_cs', 'visitor_gidp', 'visitor_ci', 'visitor_lob', 'visitor_pitchers_used', 'visitor_individual_er', 'visitor_team_er', 'visitor_wp', 'visitor_balks', 'visitor_putouts', 'visitor_assists', 'visitor_errors', 'visitor_passed_balls', 'visitor_double_plays', 'visitor_triple_plays', 'home_ab', 'home_h', 'home_2b', 'home_3b', 'home_hr', 'home_rbi', 'home_sh', 'home_sf', 'home_hbp', 'home_bb', 'home_ibb', 'home_so', 'home_sb', 'home_cs', 'home_gidp', 'home_ci', 'home_lob', 'home_pitchers_used', 'home_individual_er', 'home_team_

In [9]:
"""Create a variable to identify the winner of the game [using home_score and visitor_score]"""
Teams["hwin"] = np.where(Teams["home_score"] > Teams["visitor_score"], 1, 0)
Teams["awin"] = np.where(Teams["home_score"] < Teams["visitor_score"], 1, 0)
Teams["tie"] = np.where(Teams["home_score"] == Teams["visitor_score"], 1, 0)

In [10]:
"""Create a variable that defines the season for each game [extract from the date of the game]""" 
Teams["year"] = Teams["date"].astype(str).str[0:4]  #date format is: YYYYMMDD --> change to string and get 1st 4 characters

In [11]:
display(Teams)

date  game_number  day visitor visitor_lg  visitor_game_number  \
0      19990404            0  Sun     COL         NL                    1   
1      19990405            0  Mon     TBA         AL                    1   
2      19990405            0  Mon     BOS         AL                    1   
3      19990405            0  Mon     NYA         AL                    1   
4      19990405            0  Mon     CHA         AL                    1   
...         ...          ...  ...     ...        ...                  ...   
12137  20030928            0  Sun     NYN         NL                  161   
12138  20030928            0  Sun     MIL         NL                  162   
12139  20030928            0  Sun     ATL         NL                  162   
12140  20030928            0  Sun     COL         NL                  162   
12141  20030928            0  Sun     LAN         NL                  162   

      home home_lg  home_game_number  visitor_score  ...  home_batter_8_pos  \
0      SDN      NL                 1              8  ...                  6   
1      BAL      AL                 1              7  ...                  2   
2      KCA      AL                 1              5  ...                  6   
3      OAK      AL                 1              3  ...                  2   
4      SEA      AL                 1              8  ...                  5   
...    ...     ...               ...            ...  ...                ...   
12137  FLO      NL               162              0  ...                  2   
12138  HOU      NL               162              5  ...                  6   
12139  PHI      NL               162              5  ...                  5   
12140  SDN      NL               162             10  ...                  2   
12141  SFN      NL               161              3  ...                  2   

       home_batter_9_id home_batter_9_name home_batter_9_pos  additional_info  \
0              ashba002         Andy Ashby                 1              NaN   
1              reboj001      Jeff Reboulet                 4              NaN   
2              feblc001      Carlos Febles                 4              NaN   
3              tejam001      Miguel Tejada                 6              NaN   
4              wilsd001         Dan Wilson                 2              NaN   
...                 ...                ...               ...              ...   
12137          willd003   Dontrelle Willis                 1              NaN   
12138          saark001      Kirk Saarloos                 1              NaN   
12139          millk004     Kevin Millwood                 1              NaN   
12140          peavj001         Jake Peavy                 1              NaN   
12141          corrk001      Kevin Correia                 1              NaN   

      acquisition hwin  awin  tie  year  
0               Y    0     1    0  1999  
1               Y    1     0    0  1999  
2               Y    0     1    0  1999  
3               Y    1     0    0  1999  
4               Y    0     1    0  1999  
...           ...  ...   ...  ...   ...  
12137           Y    1     0    0  2003  
12138           Y    1     0    0  2003  
12139           Y    0     1    0  2003  
12140           Y    0     1    0  2003  
12141           Y    1     0    0  2003  

[12142 rows x 165 columns]

**Aggregate the required statistics:**
* ab - at bat
* h - hits
* 1b - singles [= h - 2b - 3b - hr]
* 2b - double (second base)
* 3b - triple (third base)
* hr - homerun
* sf - sacrifice flies
* bb - walk
* hbp - hit by pitch

### Create 2 databases:
* For our analysis we need both: slugging & on-base percentage of a team across a season AND slugging & on-base percentage against that team (of the opponents) across the season. 
* To aggeregate the data we need to generate all of the home and away records for each team. So we will create 2 aggregations and then merge them to create a single set of records for each team in the season.

In [12]:
# Measure & Aggregate performance of each team while playing at home
Teamshome = Teams.groupby(["home", "year"])["visitor_ab", "visitor_h", "visitor_2b", "visitor_3b", "visitor_hr", 
                                              "visitor_sf", "visitor_hbp", "visitor_bb", "home_ab", "home_h", "home_2b", 
                                              "home_3b", "home_hr", "home_sf", "home_hbp", "home_bb", "hwin", "awin"].sum()
Teamshome = Teamshome.reset_index()
Teamshome = Teamshome.rename(columns = {"home":"Team"})
Teamshome

<ipython-input-12-af12b86e7d01>:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  Teamshome = Teams.groupby(["home", "year"])["visitor_ab", "visitor_h", "visitor_2b", "visitor_3b", "visitor_hr",


Team  year  visitor_ab  visitor_h  visitor_2b  visitor_3b  visitor_hr  \
0    ANA  1999        2832        762         151          10         105   
1    ANA  2000        2867        800         144           8         113   
2    ANA  2001        2873        772         165           6          86   
3    ANA  2002        2799        685         132           5          70   
4    ANA  2003        2868        743         142           9          84   
..   ...   ...         ...        ...         ...         ...         ...   
145  TOR  1999        2891        817         190          11          93   
146  TOR  2000        2910        819         182          11          92   
147  TOR  2001        2911        800         207          17          84   
148  TOR  2002        2838        749         171          10          89   
149  TOR  2003        2930        827         172          19         107   

     visitor_sf  visitor_hbp  visitor_bb  home_ab  home_h  home_2b  home_3b  \
0            39           27         322     2666     669      118       13   
1            24           15         320     2742     788      146       11   
2            20           39         259     2736     731      122       12   
3            29           22         260     2757     769      155       21   
4            17           39         245     2685     717      135       17   
..          ...          ...         ...      ...     ...      ...      ...   
145          31           25         289     2731     765      173        6   
146          19           30         279     2766     768      178        9   
147          25           46         264     2809     757      146       14   
148          28           37         288     2729     734      164       20   
149          20           33         244     2763     779      191       16   

     home_hr  home_sf  home_hbp  home_bb  hwin  awin  
0         74       21        23      267    37    44  
1        130       23        24      315    46    35  
2         86       29        30      251    39    42  
3         71       27        34      228    54    27  
4         68       22        27      226    45    37  
..       ...      ...       ...      ...   ...   ...  
145       96       21        44      283    40    41  
146      134       12        33      255    45    36  
147       94       31        37      256    40    42  
148      102       24        30      258    42    39  
149       94       28        40      284    41    40  

[150 rows x 20 columns]

### For home team, you need to group first by team name, and then by the season. This is because the team will play several home games every year

In [13]:
# Measure & Aggregate performance of each team while playing away
Teamsaway = Teams.groupby(["year", "visitor"])["visitor_ab", "visitor_h", "visitor_2b", "visitor_3b", "visitor_hr", 
                                              "visitor_sf", "visitor_hbp", "visitor_bb", "home_ab", "home_h", "home_2b", 
                                              "home_3b", "home_hr", "home_sf", "home_hbp", "home_bb", "hwin", "awin"].sum()
Teamsaway = Teamsaway.reset_index()
Teamsaway = Teamsaway.rename(columns = {"visitor":"Team"})
Teamsaway

<ipython-input-13-dbc35c0de7d4>:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  Teamsaway = Teams.groupby(["year", "visitor"])["visitor_ab", "visitor_h", "visitor_2b", "visitor_3b", "visitor_hr",


year Team  visitor_ab  visitor_h  visitor_2b  visitor_3b  visitor_hr  \
0    1999  ANA        2828        735         130           9          84   
1    1999  ARI        2928        812         147          15         115   
2    1999  ATL        2901        777         151          14         111   
3    1999  BAL        2905        810         163          12         105   
4    1999  BOS        2857        758         157          22          96   
..    ...  ...         ...        ...         ...         ...         ...   
145  2003  SFN        2752        689         147          10          98   
146  2003  SLN        2914        781         159          21         111   
147  2003  TBA        2859        749         140          17          81   
148  2003  TEX        2877        708         133          13          99   
149  2003  TOR        2898        801         166          17          96   

     visitor_sf  visitor_hbp  visitor_bb  home_ab  home_h  home_2b  home_3b  \
0            21           20         244     2645     710      134       15   
1            29           31         303     2751     694      123       12   
2            19           31         322     2772     724      131        9   
3            23           25         300     2638     708      157       19   
4            27           25         292     2704     674      134       10   
..          ...          ...         ...      ...     ...      ...      ...   
145          18           26         287     2590     657      124       12   
146          30           34         300     2831     815      180       16   
147          29           27         188     2663     719      163       17   
148          15           41         242     2719     770      154       17   
149          28           50         262     2730     733      128       23   

     home_hr  home_sf  home_hbp  home_bb  hwin  awin  
0         72       26        29      302    48    33  
1         92       14        22      287    33    48  
2         67       22        16      286    34    47  
3         97       22        19      318    44    37  
4         88       23        23      243    36    45  
..       ...      ...       ...      ...   ...   ...  
145       75       26        23      285    37    43  
146      117       27        37      276    44    37  
147      100       30        32      317    54    27  
148      103       34        24      319    53    28  
149       77       18        24      241    36    45  

[150 rows x 20 columns]

### For away teams, we group first by year and then by the team.

In [14]:
"""Merge the 2 datasets"""
Teams2 = pd.merge(Teamshome, Teamsaway, on = ["Team", "year"])
pd.set_option('display.max_columns', 50)
display(Teams2)

Team  year  visitor_ab_x  visitor_h_x  visitor_2b_x  visitor_3b_x  \
0    ANA  1999          2832          762           151            10   
1    ANA  2000          2867          800           144             8   
2    ANA  2001          2873          772           165             6   
3    ANA  2002          2799          685           132             5   
4    ANA  2003          2868          743           142             9   
..   ...   ...           ...          ...           ...           ...   
145  TOR  1999          2891          817           190            11   
146  TOR  2000          2910          819           182            11   
147  TOR  2001          2911          800           207            17   
148  TOR  2002          2838          749           171            10   
149  TOR  2003          2930          827           172            19   

     visitor_hr_x  visitor_sf_x  visitor_hbp_x  visitor_bb_x  home_ab_x  \
0             105            39             27           322       2666   
1             113            24             15           320       2742   
2              86            20             39           259       2736   
3              70            29             22           260       2757   
4              84            17             39           245       2685   
..            ...           ...            ...           ...        ...   
145            93            31             25           289       2731   
146            92            19             30           279       2766   
147            84            25             46           264       2809   
148            89            28             37           288       2729   
149           107            20             33           244       2763   

     home_h_x  home_2b_x  home_3b_x  home_hr_x  home_sf_x  home_hbp_x  \
0         669        118         13         74         21          23   
1         788        146         11        130         23          24   
2         731        122         12         86         29          30   
3         769        155         21         71         27          34   
4         717        135         17         68         22          27   
..        ...        ...        ...        ...        ...         ...   
145       765        173          6         96         21          44   
146       768        178          9        134         12          33   
147       757        146         14         94         31          37   
148       734        164         20        102         24          30   
149       779        191         16         94         28          40   

     home_bb_x  hwin_x  awin_x  visitor_ab_y  visitor_h_y  visitor_2b_y  \
0          267      37      44          2828          735           130   
1          315      46      35          2886          786           163   
2          251      39      42          2815          716           153   
3          228      54      27          2921          834           178   
4          226      45      37          2802          756           141   
..         ...     ...     ...           ...          ...           ...   
145        283      40      41          2911          815           164   
146        255      45      36          2911          794           150   
147        256      40      42          2854          732           141   
148        258      42      39          2852          723           141   
149        284      41      40          2898          801           166   

     visitor_3b_y  visitor_hr_y  visitor_sf_y  visitor_hbp_y  visitor_bb_y  \
0               9            84            21             20           244   
1              23           106            20             23           293   
2              14            72            24             47           243   
3              11            81            37             40           234   
4              16            82            28             

### Merged Dataset
* The dataframe contains the records for each team, in each season.
* All the '_x'  variables relate to the team as home team
* All the '_y' variables relate to the team as away team


* For Team ANA:
    * visitor_ab_x = at bats of the opposite team when ANA played at home
    * home_ab_x = at bats of ANA when they played at home
    * visitor_ab_y = at bats of ANA (visitor team) when ANA played away
    * home_ab_y = at bats of opposite team (home team) when ANA played away 

In [15]:
"""Total number of wins for each team in each season"""
Teams2["wins"] = Teams2["hwin_x"] + Teams2["awin_y"] #home wins when playing as hometeam + away wins when playing as away team

### Calculate the slugging percentage and on-base percentage

In [16]:
"""On-base percentage [OBP] = (h + bb + hbp) / (ab + bb + hpb + sf) for and against each team"""

#First calculate OBP for a team [OBPFOR]: OBP of team while playing at home + OBP of team while playing as away team
Teams2["OBPFOR"] = ((Teams2['home_h_x'] + Teams2['visitor_h_y'] + Teams2['home_bb_x'] + Teams2['visitor_bb_y']
                    + Teams2['home_hbp_x'] + Teams2['visitor_hbp_y']) / (Teams2['home_ab_x'] + Teams2['visitor_ab_y']
                    + Teams2['home_bb_x'] + Teams2['visitor_bb_y'] + Teams2['home_hbp_x'] + Teams2['visitor_hbp_y']
                    + Teams2['home_sf_x'] + Teams2['visitor_sf_y']))

In [17]:
#Calculate OBP against a team [OBPAGN]: OBP against a team while they play at home + OBP against the team when playing away
Teams2["OBPAGN"] = ((Teams2['home_h_y'] + Teams2['visitor_h_x'] + Teams2['home_bb_y'] + Teams2['visitor_bb_x']
                    + Teams2['home_hbp_y'] + Teams2['visitor_hbp_x']) / (Teams2['home_ab_y'] + Teams2['visitor_ab_x']
                    + Teams2['home_bb_y'] + Teams2['visitor_bb_x'] + Teams2['home_hbp_y'] + Teams2['visitor_hbp_x']
                    + Teams2['home_sf_y'] + Teams2['visitor_sf_x']))

In [18]:
"""Slogging percentage [SLG] = ((h - 2b - 3b - hr) + 2(2b) + 3(3b) + 4(hr)) / ab"""

#First calculate SLG for a team [SLGFOR]: SLG of team while playing at home + SLG of team while playing as away team
Teams2["SLGFOR"] = ((Teams2["home_h_x"] + Teams2["visitor_h_y"] - 
                     (Teams2["home_2b_x"] + Teams2["visitor_2b_y"]) - 
                     (Teams2["home_3b_x"] + Teams2["visitor_3b_y"]) -
                     (Teams2["home_hr_x"] + Teams2["visitor_hr_y"]) + 
                     2*Teams2["home_2b_x"] + 2*Teams2["visitor_2b_y"] + 
                     3*Teams2["home_3b_x"] + 3*Teams2["visitor_3b_y"] +
                     4*Teams2["home_hr_x"] + 4*Teams2["visitor_hr_y"]) /
                     (Teams2["home_ab_x"] + Teams2["visitor_ab_y"]))

In [19]:
#Calculate SLG against a team [SLGAGN]: SLG against a team while they play at home + SLG against the team when playing away
Teams2["SLGAGN"] = ((Teams2["home_h_y"] + Teams2["visitor_h_x"] - 
                     (Teams2["home_2b_y"] + Teams2["visitor_2b_x"]) - 
                     (Teams2["home_3b_y"] + Teams2["visitor_3b_x"]) -
                     (Teams2["home_hr_y"] + Teams2["visitor_hr_x"]) + 
                     2*Teams2["home_2b_y"] + 2*Teams2["visitor_2b_x"] + 
                     3*Teams2["home_3b_y"] + 3*Teams2["visitor_3b_x"] +
                     4*Teams2["home_hr_y"] + 4*Teams2["visitor_hr_x"]) /
                     (Teams2["home_ab_y"] + Teams2["visitor_ab_x"]))

In [20]:
"""Create a new dataset, with only the required columns"""
Teams3 = Teams2[["year", "Team", "wins", "OBPFOR", "OBPAGN", "SLGFOR", "SLGAGN"]]
Teams3

year Team  wins    OBPFOR    OBPAGN    SLGFOR    SLGAGN
0    1999  ANA    70  0.321511  0.345869  0.394976  0.426876
1    2000  ANA    82  0.352355  0.350943  0.472459  0.452844
2    2001  ANA    75  0.326802  0.331493  0.404972  0.412190
3    2002  ANA    99  0.340714  0.313509  0.432547  0.392078
4    2003  ANA    77  0.330367  0.326923  0.412794  0.420824
..    ...  ...   ...       ...       ...       ...       ...
145  1999  TOR    84  0.352310  0.349186  0.457462  0.451099
146  2000  TOR    83  0.341115  0.353545  0.469262  0.455812
147  2001  TOR    80  0.325280  0.338607  0.429631  0.434944
148  2002  TOR    78  0.327056  0.343869  0.429851  0.430772
149  2003  TOR    86  0.348812  0.336859  0.454513  0.440989

[150 rows x 7 columns]

### Calculate the win percentage for each team
* For win percentage, we need the number of games played by each team: use groupby on Teams df on year and home --> .count()

In [21]:
TeamsGh = Teams.groupby(['year', 'home'])['hwin'].count()   #home games played by each team
TeamsGh = TeamsGh.reset_index()
TeamsGh = TeamsGh.rename(columns = {"home":"Team"})

TeamsGa = Teams.groupby(['year', 'visitor'])['awin'].count()  #away games played by each team
TeamsGa = TeamsGa.reset_index()
TeamsGa = TeamsGa.rename(columns = {"visitor":"Team"})

TeamsG = pd.merge(TeamsGh, TeamsGa, on = ["Team", "year"])
TeamsG["Games"] = TeamsG["hwin"] + TeamsG["awin"]
TeamsG = TeamsG[["year", "Team", "Games"]]  #total games played by each team
TeamsG

year Team  Games
0    1999  ANA    162
1    1999  ARI    162
2    1999  ATL    162
3    1999  BAL    162
4    1999  BOS    162
..    ...  ...    ...
145  2003  SFN    161
146  2003  SLN    162
147  2003  TBA    162
148  2003  TEX    162
149  2003  TOR    162

[150 rows x 3 columns]

In [22]:
Teams3 = pd.merge(Teams3, TeamsG, on = ["Team", "year"])
Teams3.head(10)

year Team  wins    OBPFOR    OBPAGN    SLGFOR    SLGAGN  Games
0  1999  ANA    70  0.321511  0.345869  0.394976  0.426876    162
1  2000  ANA    82  0.352355  0.350943  0.472459  0.452844    162
2  2001  ANA    75  0.326802  0.331493  0.404972  0.412190    162
3  2002  ANA    99  0.340714  0.313509  0.432547  0.392078    162
4  2003  ANA    77  0.330367  0.326923  0.412794  0.420824    162
5  1999  ARI   100  0.346553  0.319864  0.458643  0.402156    162
6  2000  ARI    85  0.333387  0.325830  0.429347  0.424077    162
7  2001  ARI    92  0.340717  0.310654  0.441823  0.404209    162
8  2002  ARI    98  0.346019  0.304882  0.423203  0.396843    162
9  2003  ARI    84  0.329622  0.321673  0.416517  0.388436    162

In [23]:
"""Calculate Win Percentage"""
Teams3["wpc"] = Teams3["wins"] / Teams3["Games"]
Teams3

year Team  wins    OBPFOR    OBPAGN    SLGFOR    SLGAGN  Games       wpc
0    1999  ANA    70  0.321511  0.345869  0.394976  0.426876    162  0.432099
1    2000  ANA    82  0.352355  0.350943  0.472459  0.452844    162  0.506173
2    2001  ANA    75  0.326802  0.331493  0.404972  0.412190    162  0.462963
3    2002  ANA    99  0.340714  0.313509  0.432547  0.392078    162  0.611111
4    2003  ANA    77  0.330367  0.326923  0.412794  0.420824    162  0.475309
..    ...  ...   ...       ...       ...       ...       ...    ...       ...
145  1999  TOR    84  0.352310  0.349186  0.457462  0.451099    162  0.518519
146  2000  TOR    83  0.341115  0.353545  0.469262  0.455812    162  0.512346
147  2001  TOR    80  0.325280  0.338607  0.429631  0.434944    162  0.493827
148  2002  TOR    78  0.327056  0.343869  0.429851  0.430772    162  0.481481
149  2003  TOR    86  0.348812  0.336859  0.454513  0.440989    162  0.530864

[150 rows x 9 columns]

### Running the Regression

In [24]:
import statsmodels.formula.api as smf 

In [25]:
winOBP_lm = smf.ols(formula = "wpc ~ OBPFOR + OBPAGN", data = Teams3).fit()
winOBP_lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    wpc   R-squared:                       0.826
Model:                            OLS   Adj. R-squared:                  0.823
Method:                 Least Squares   F-statistic:                     348.5
Date:                Sun, 12 Dec 2021   Prob (F-statistic):           1.62e-56
Time:                        14:06:14   Log-Likelihood:                 301.40
No. Observations:                 150   AIC:                            -596.8
Df Residuals:                     147   BIC:                            -587.8
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5030      0.087      5.776      0.000       0.331       0.675
OBPFOR         3.3009      0.189     17.459      0.000       2.927       3.675
OBPAGN        -3.3105      0.171    -19.341      0.000      -3.649      -2.972
==============================================================================
Omnibus:                        1.861   Durbin-Watson:                   1.889
Prob(Omnibus):                  0.394   Jarque-Bera (JB):                1.460
Skew:                          -0.013   Prob(JB):                        0.482
Kurtosis:                       2.517   Cond. No.                         85.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [26]:
winSLG_lm = smf.ols(formula = "wpc ~ SLGFOR + SLGAGN", data = Teams3).fit()
winSLG_lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    wpc   R-squared:                       0.786
Model:                            OLS   Adj. R-squared:                  0.783
Method:                 Least Squares   F-statistic:                     270.4
Date:                Sun, 12 Dec 2021   Prob (F-statistic):           5.57e-50
Time:                        14:06:14   Log-Likelihood:                 286.04
No. Observations:                 150   AIC:                            -566.1
Df Residuals:                     147   BIC:                            -557.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5904      0.065      9.015      0.000       0.461       0.720
SLGFOR         1.7552      0.116     15.134      0.000       1.526       1.984
SLGAGN        -1.9676      0.107    -18.456      0.000      -2.178      -1.757
==============================================================================
Omnibus:                        0.471   Durbin-Watson:                   1.716
Prob(Omnibus):                  0.790   Jarque-Bera (JB):                0.565
Skew:                          -0.127   Prob(JB):                        0.754
Kurtosis:                       2.838   Cond. No.                         50.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [27]:
winOBPSLG_lm = smf.ols(formula = "wpc ~ OBPFOR + OBPAGN + SLGFOR + SLGAGN", data = Teams3).fit()
winOBPSLG_lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    wpc   R-squared:                       0.885
Model:                            OLS   Adj. R-squared:                  0.882
Method:                 Least Squares   F-statistic:                     279.0
Date:                Sun, 12 Dec 2021   Prob (F-statistic):           5.15e-67
Time:                        14:06:14   Log-Likelihood:                 332.53
No. Observations:                 150   AIC:                            -655.1
Df Residuals:                     145   BIC:                            -640.0
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5052      0.073      6.930      0.000       0.361       0.649
OBPFOR         2.1215      0.255      8.331      0.000       1.618       2.625
OBPAGN        -1.9156      0.256     -7.489      0.000      -2.421      -1.410
SLGFOR         0.8143      0.141      5.775      0.000       0.536       1.093
SLGAGN        -0.9897      0.144     -6.868      0.000      -1.275      -0.705
==============================================================================
Omnibus:                        1.334   Durbin-Watson:                   2.059
Prob(Omnibus):                  0.513   Jarque-Bera (JB):                1.426
Skew:                          -0.196   Prob(JB):                        0.490
Kurtosis:                       2.727   Cond. No.                         165.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [28]:
winOBPSLGR_lm = smf.ols(formula = "wpc ~ I(OBPFOR-OBPAGN) + I(SLGFOR-SLGAGN)", data = Teams3).fit()
winOBPSLGR_lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    wpc   R-squared:                       0.884
Model:                            OLS   Adj. R-squared:                  0.883
Method:                 Least Squares   F-statistic:                     561.8
Date:                Sun, 12 Dec 2021   Prob (F-statistic):           1.43e-69
Time:                        14:06:14   Log-Likelihood:                 332.07
No. Observations:                 150   AIC:                            -658.1
Df Residuals:                     147   BIC:                            -649.1
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept              0.4997      0.002    229.103      0.000       0.495       0.504
I(OBPFOR - OBPAGN)     2.0327      0.180     11.318      0.000       1.678       2.388
I(SLGFOR - SLGAGN)     0.8992      0.104      8.618      0.000       0.693       1.105
==============================================================================
Omnibus:                        1.506   Durbin-Watson:                   2.048
Prob(Omnibus):                  0.471   Jarque-Bera (JB):                1.573
Skew:                          -0.229   Prob(JB):                        0.455
Kurtosis:                       2.796   Cond. No.                         92.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [29]:
"""Combine all the results together"""
from statsmodels.iolib.summary2 import summary_col
Header = ["1","2","3","4"]
Table_1 = summary_col([winOBP_lm, winSLG_lm, winOBPSLG_lm, winOBPSLGR_lm], 
                      regressor_order = ['Intercept', 'OBPFOR', 'OBPAGN', 'SLGFOR', 'SLGAGN'], model_names = Header)
print(Table_1)


                      1        2        3        4    
------------------------------------------------------
Intercept          0.5030   0.5904   0.5052   0.4997  
                   (0.0871) (0.0655) (0.0729) (0.0022)
OBPFOR             3.3009            2.1215           
                   (0.1891)          (0.2546)         
OBPAGN             -3.3105           -1.9156          
                   (0.1712)          (0.2558)         
SLGFOR                      1.7552   0.8143           
                            (0.1160) (0.1410)         
SLGAGN                      -1.9676  -0.9897          
                            (0.1066) (0.1441)         
I(SLGFOR - SLGAGN)                            0.8992  
                                              (0.1043)
I(OBPFOR - OBPAGN)                            2.0327  
                                              (0.1796)
R-squared          0.8258   0.7863   0.8850   0.8843  
R-squared Adj.     0.8235   0.7834   0.8818   0.8827  
Standard 

#### Interpretations of the result:
* The coefficient value of OBPFOR (3.3009) is greater than that of SLGFOR (1.7552) --> **The onbase percentage, is more important than the slugging percentage, when determing the win percentage of a team**
* This result therefore shows that ideally, the teams should have given a significant consideration to OBP when determing a player's salary!